frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts

```diff
@@ -223,14 +229,87 @@ export const getEdgesFromNodes = (nodes: PipelineMixedNodeModel[]): PipelineEdge
     }),
   ).filter((edgeList) => !!edgeList);
 
+export const getFinallyTaskHeight = (allTasksLength: number, disableBuilder: boolean): number => {
+  const FINALLY_TASK_NODE_HEIGHT = NODE_HEIGHT + FINALLY_NODE_VERTICAL_SPACING;
+  return (
+    NODE_HEIGHT +
+    (allTasksLength - 1) * FINALLY_TASK_NODE_HEIGHT +
+    (!disableBuilder ? NODE_HEIGHT : 0) +
+    FINALLY_NODE_PADDING * 2
+  );
+};
+
+export const getLastRegularTasks = (nodes: PipelineMixedNodeModel[]): string[] => {
+  const regularTasks = nodes;
+  const runAfters: string[] = regularTasks.reduce((acc, regularTask) => {
+    const {
+      data: {
+        task: { runAfter },
+      },
+    } = regularTask;
+    if (runAfter && runAfter.length) {
+      runAfter.forEach((ra) => {
+        if (!acc.includes(ra)) {
+          acc.push(ra);
+        }
+      });
+    }
+    return acc;
+  }, []);
+
+  return _.difference(
+    regularTasks.map((n) => n.id),
+    runAfters,
+  );
+};
+
+export const finallyTasksToNodes = (
+  nodes: PipelineMixedNodeModel[],
+  pipeline?: PipelineKind,
+  pipelineRun?: PipelineRunKind,
+): PipelineMixedNodeModel[] => {
+  if (!pipeline.spec?.finally) {
+    return nodes;
+  }
+  const finallyTasks = pipeline.spec.finally;
+  const regularRunAfters = getLastRegularTasks(nodes);
+  const name = 'finally-node';
+  const finallyGroupNode = createFinallyNode(getFinallyTaskHeight(finallyTasks.length, true))(
+    name,
+    {
+      disableBuilder: true,
+      isFinallyTask: true,
+      pipeline,
+      pipelineRun,
+      task: {
+        name,
+        runAfter: regularRunAfters,
+        finallyTasks: finallyTasks.map((ft) => ({
+          ...ft,
+          disableTooltip: false,
+        })),
+      },
+    },
+  );
+  return [finallyGroupNode];
+};
+
 export const getTopologyNodesEdges = (
   pipeline: PipelineKind,
   pipelineRun?: PipelineRunKind,
 ): { nodes: PipelineMixedNodeModel[]; edges: PipelineEdgeModel[] } => {
   const taskList: PipelineVisualizationTaskItem[] = _.flatten(
     getPipelineTasks(pipeline, pipelineRun),
   );
-  const nodes: PipelineMixedNodeModel[] = tasksToNodes(taskList, pipeline, pipelineRun);
+  const taskNodes: PipelineMixedNodeModel[] = tasksToNodes(taskList, pipeline, pipelineRun);
+
+  const finallyTaskNodes: PipelineMixedNodeModel[] = finallyTasksToNodes(
+    taskNodes,
+    pipeline,
+    pipelineRun,
+  );
+
+  const nodes = [...taskNodes, ...finallyTaskNodes];
```

Comment:
Also you could have just done a singular node and placed it as the last item... you seemed to have wrapped it in an array for some reason I'm not seeing...

Reply:
Should be singular node, updated